File: threads02/java/ex02/src/model/Dados.java
package model;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public class Dados {
    private static final int NUM_PARTES = 2;

    /**
     * método que carrega os nomes do arquivo
     * @param caminho caminho do arquivo
     * @return lista de nomes
     * @throws IOException caso o arquivo não seja encontrado
     */
    public List<String> carregarNomes(String caminho) throws IOException {
        return Files.readAllLines(Path.of(caminho));
    }

    /**
     * método que divide a lista de nomes em partes
     * @param nomes lista de nomes
     * @return lista de listas de nomes
     */
    public List<List<String>> dividirEmPartes(List<String> nomes) {
        if (nomes == null || nomes.isEmpty()) {
            throw new IllegalArgumentException("A lista de nomes não pode ser nula ou vazia");
        }

        int tamanhoBase = nomes.size() / NUM_PARTES;
        int resto = nomes.size() % NUM_PARTES;

        List<List<String>> partes = new ArrayList<>();

        int inicio = 0;

        for (int i = 0; i < NUM_PARTES; i++) {

            int tamanhoParte = tamanhoBase + (i < resto ? 1 : 0);
            int fim = inicio + tamanhoParte;

            partes.add(
                new ArrayList<>(nomes.subList(inicio, fim))
            );

            inicio = fim;
        }

        return partes;
    }
}

File: threads02/csharp/ex02/Model/Dados.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace ex02.Model;

public class Dados
{
    private const int NumPartes = 2;

    public List<string> CarregarNomes(string caminho)
    {
        if (string.IsNullOrWhiteSpace(caminho))
        {
            throw new ArgumentException("O caminho do arquivo não pode ser nulo ou vazio");
        }

        return File.ReadAllLines(caminho).ToList();
    }

    public List<List<string>> DividirEmPartes(List<string> nomes)
    {
        if (nomes == null || nomes.Count == 0)
        {
            throw new ArgumentException("A lista de nomes não pode ser nula ou vazia.");
        }

        List<List<string>> partes = new();

        int tamanhoBase = nomes.Count / NumPartes;
        int resto = nomes.Count % NumPartes;
        int inicio = 0;

        for (int i = 0; i < NumPartes; i++)
        {
            int tamanhoParte = tamanhoBase + (i < resto ? 1 : 0);

            partes.Add(nomes.GetRange(inicio, tamanhoParte));

            inicio += tamanhoParte;
        }

        return partes;
    }
}
